pgvectorscale/src/util/chain.rs

```diff
@@ -0,0 +1,280 @@
+//! This module defines the `ChainTape` data structure, which is used to store large data items that
+//! are too big to fit in a single page.  See `Tape` for a simpler version that assumes each data
+//! item fits in a single page.
+//!
+//! All page entries begin with a header that contains a block number of the next page in the chain,
+//! if applicable.  If there is a next page, the entry continues (at position 1) on that page.
+//!
+//! The implementation supports an append-only sequence of writes via `ChainTapeWriter` and reads
+//! via `ChainTapeReader`.  The writer returns an `ItemPointer` that can be used to read the data
+//! back.  Reads are done via an iterator that returns `ReadableBuffer` objects for the segments
+//! of the data.
+
+use anyhow::Result;
+
+use pgrx::{
+    pg_sys::{BlockNumber, InvalidBlockNumber},
+    PgRelation,
+};
+use rkyv::{Archive, Deserialize, Serialize};
+
+use crate::access_method::stats::{StatsNodeRead, StatsNodeWrite};
+
+use super::{
+    page::{PageType, ReadablePage, WritablePage},
+    ItemPointer, ReadableBuffer,
+};
+
+#[derive(Clone, PartialEq, Archive, Deserialize, Serialize)]
+#[archive(check_bytes)]
+struct ChainItemHeader {
+    next: BlockNumber,
+}
+
+const CHAIN_HEADER_SIZE: usize = std::mem::size_of::<ArchivedChainItemHeader>();
+
+// Empirically-measured slop factor for how much `pg_sys::PageGetFreeSpace` can
+// overestimate the free space in a page in our usage patterns.
+const PG_SLOP_SIZE: usize = 4;
+
+pub struct ChainTapeWriter<'a, S: StatsNodeWrite> {
+    page_type: PageType,
+    index: &'a PgRelation,
+    current: BlockNumber,
+    stats: &'a mut S,
+}
+
+impl<'a, S: StatsNodeWrite> ChainTapeWriter<'a, S> {
+    /// Create a ChainingTape that starts writing on a new page.
+    pub fn new(index: &'a PgRelation, page_type: PageType, stats: &'a mut S) -> Self {
+        assert!(page_type.is_chained());
+        let page = WritablePage::new(index, page_type);
+        let block_number = page.get_block_number();
+        page.commit();
+        Self {
+            page_type,
+            index,
+            current: block_number,
+            stats,
+        }
+    }
+
+    /// Write chained data to the tape, returning an `ItemPointer` to the start of the data.
+    pub fn write(&mut self, mut data: &[u8]) -> Result<super::ItemPointer> {
```

Comment:
This function returns a `anyhow::Result` but always returns `Ok(...)` - why wrap it in a Result? Also why are you using `anyhow` here?